### PR TITLE
Allows limiting of filtered changes requests

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

When CouchDB version is gte 2.3.0, uses the `limit` param when requesting `_changes`, to allow a batched filtered replication.

medic/medic-webapp#5076

# Review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- [ ] Documented: Announced in Changes.md in plain English. Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by dianabarsan** [comment]
This fails miserably, because it needs the `ServerChecks` update found in https://github.com/medic/medic-webapp/pull/5112 .

**Comment by dianabarsan** [comment]
Hey @garethbowen.
I know the tests don't pass (because of the missing shared lib), but could you have a look to get an idea of what I'm trying to do? 
In the end, I kept processing live changes for all feeds, since there's no way of reliably checking how up to date you are, and if you are up to date you will miss those changes without that code. 

I also have to update this PR to depend on the new version of @medic/server-checks, when it will exist. 

Thanks in advance!

File: tests/e2e/api/controllers/_changes.spec.js [hunk 1]
```diff
@@ -681,8 +771,8 @@ describe('changes handler', () => {
                         utils.saveDocs(allowedSteve),
                       ]))
                       .then(resolve);
-                  });
-                }, 1000)
+                  }, 1000);
```
Comment (dianabarsan) [inline tests/e2e/api/controllers/_changes.spec.js]: awkward :grimacing:

File: api/tests/mocha/controllers/changes.js [hunk 2]
```diff
@@ -32,6 +35,7 @@ describe('Changes controller', () => {
   });
 
   beforeEach(() => {
+    realSetTimeout = setTimeout;
```
Comment (garethbowen) [inline api/tests/mocha/controllers/changes.js]: Why not just inline `setTimeout`?

File: tests/e2e/api/controllers/_changes.spec.js (hunk 1 above)
Comment (garethbowen) [reply tests/e2e/api/controllers/_changes.spec.js]: Hah.

File: tests/e2e/api/controllers/_changes.spec.js [hunk 3]
```diff
@@ -423,12 +436,89 @@ describe('changes handler', () => {
...
+    });
+
     beforeEach(done => getCurrentSeq().then(done));
 
-    const bobsIds = [...DEFAULT_EXPECTED],
-          stevesIds = [...DEFAULT_EXPECTED];
+    it('should successfully fully replicate in batches', () => {
+      if (!shouldBatchChangesRequests) {
+        return;
+      }
```
Comment (garethbowen) [inline tests/e2e/api/controllers/_changes.spec.js]: I think it'd be better to move the `if` block outside the test so the test doesn't get reported as passing if it's not even run, eg:

```js
if (shouldBatchChangesRequests) {
  it('should successfully fully replicate in batches', () => {
    ...
```

This applies to the tests below too...

File: tests/e2e/api/controllers/_changes.spec.js [hunk 4]
```diff
@@ -4,7 +4,8 @@ const _ = require('underscore'),
       http = require('http'),
       querystring = require('querystring'),
       constants = require('../../../constants'),
-      auth = require('../../../auth')();
+      auth = require('../../../auth')(),
+      semver = require('semver');
```
Comment (garethbowen) [inline tests/e2e/api/controllers/_changes.spec.js]: I can't find where `semver` is listed in a package.json so I think this is just a transitive dependency currently so we can't rely on it being there forever. Add it to the appropriate package.json so it doesn't disappear!

File: api/tests/mocha/controllers/changes.js (hunk 2 above)
Comment (dianabarsan) [reply api/tests/mocha/controllers/changes.js]: hm? I need to save the proper `setTimeout` function (which gets overriden by `useFakeTimers` on next line) to use as `nextTick` (the promise chain became unusable). 
Do you have another suggestion?

Comment (dianabarsan) [reply api/tests/mocha/controllers/changes.js]: Nevermind ... it's too late and I can't think straight :) Thanks!

Comment (garethbowen) [reply api/tests/mocha/controllers/changes.js]: Ahh, it's overridden. That makes sense.

File: api/src/controllers/changes.js [hunk 5]
```diff
@@ -346,17 +344,50 @@ const initContinuousFeed = since => {
...
+    return idx === -1 || parseInt(b[idx]) > parseInt(a[idx]);
+  };
+
+  // Prior to version 2.3.0, CouchDB had a bug where requesting _changes filtered by _doc_ids and using limit
+  // would yield an incorrect `last_seq`, resulting in overall incomplete changes.
+  const MIN_COUCH_VERSION_FOR_LIMITING_CHANGES = '2.3.0',
+        minVersion = parseVersion(MIN_COUCH_VERSION_FOR_LIMITING_CHANGES),
+        actualVersion = parseVersion(couchDbVersion);
+
+  limitChangesRequests = actualVersion ? gte(minVersion, actualVersion) : false;
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply api/src/controllers/changes.js]: I've added `semver` to API and our build dev dependencies.

File: tests/e2e/api/controllers/_changes.spec.js (hunk 3 above)
Comment (dianabarsan) [reply tests/e2e/api/controllers/_changes.spec.js]: I've ended up doing it differently, cause the test-suite list is evaluated before before/after hooks (apparently), so you can't change which tests run while running tests. 

So I merged 2 tests into 1 and conditionally run assertions on the result set.

**Comment by dianabarsan** [comment]
@garethbowen I've updated the `server-checks` dependency and now it's green.
Could you please have another look? Thank you!

File: tests/e2e/api/controllers/_changes.spec.js (hunk 3 above)
Comment (garethbowen) [reply tests/e2e/api/controllers/_changes.spec.js]: Ahh ok. Thanks.

**Pull request merged**